libnetwork_test.go

```diff
@@ -0,0 +1,349 @@
+	})
+
+	//docker create/run
+	// create - doesn't have any network interactions until the container is started
+	// run - Run a container, check the following
+	//		- etcd contains correct info
+	//		- host namespace contains the right veth with the right info
+	//    - container - container the right routes and interface
+	//	run can have the following variations -
+	//    --mac-address
+	//    --link-local-ip
+	//    --ip and --ip6
+	Describe("docker run", func() {
+		var name string
+		BeforeEach(func() {
+			name = fmt.Sprintf("run%d", rand.Uint32())
+			DockerString(fmt.Sprintf("docker network create %s -d calico --ipam-driver calico-ipam", name))
+		})
+		AfterEach(func() {
+			DockerString(fmt.Sprintf("docker network rm %s", name))
+		})
+
+		It("creates a container on a network  and checks all assertions", func() {
+			// Create a container that will just sit in the background
+			DockerString(fmt.Sprintf("docker run --net %s -tid --name %s busybox", name, name))
+
+			// Gather information for assertions
+			docker_endpoint := GetDockerEndpoint(name, name)
+			ip := docker_endpoint.IPAddress
+			mac := docker_endpoint.MacAddress
+			endpoint_id := docker_endpoint.EndpointID
+			interface_name := "cali" + endpoint_id[:mathutils.MinInt(11, len(endpoint_id))]
+
+			// Check that the endpoint is created in etcd
+			etcd_endpoint := GetEtcdString(fmt.Sprintf("/calico/v1/host/test/workload/libnetwork/libnetwork/endpoint/%s", endpoint_id))
+			Expect(etcd_endpoint).Should(MatchJSON(fmt.Sprintf(
+				`{"state":"active","name":"%s","mac":"%s","profile_ids":["%s"],"ipv4_nets":["%s/32"],"ipv6_nets":[]}`,
+				interface_name, mac, name, ip)))
+
+			// Check profile
+			tags := GetEtcdString(fmt.Sprintf("/calico/v1/policy/profile/%s/tags", name))
+			labels := GetEtcdString(fmt.Sprintf("/calico/v1/policy/profile/%s/labels", name))
+			rules := GetEtcdString(fmt.Sprintf("/calico/v1/policy/profile/%s/rules", name))
+			Expect(tags).Should(MatchJSON(fmt.Sprintf(`["%s"]`, name)))
+			Expect(labels).Should(MatchJSON("{}"))
+			Expect(rules).Should(MatchJSON(fmt.Sprintf(`{"inbound_rules": [{"action": "allow","src_tag": "%s"}],"outbound_rules":[{"action": "allow"}]}`, name)))
+
+			// Check the interface exists on the Host - it has an autoassigned
+			// mac and ip, so don't check anything!
+			DockerString(fmt.Sprintf("ip addr show %s", interface_name))
+
+			// Make sure the interface in the container exists and has the  assigned ip and mac
+			container_interface_string := DockerString(fmt.Sprintf("docker exec -i %s ip addr", name))
+			Expect(container_interface_string).Should(ContainSubstring(ip))
+			Expect(container_interface_string).Should(ContainSubstring(mac))
+
+			// Make sure the container has the routes we expect
+			routes := DockerString(fmt.Sprintf("docker exec -i %s ip route", name))
+			Expect(routes).Should(Equal("default via 169.254.1.1 dev cali0 \n169.254.1.1 dev cali0"))
```

Comment:
We should be testing the override of the container interface name via environment.